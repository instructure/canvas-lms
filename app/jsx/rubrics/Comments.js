/*
 * Copyright (C) 2018 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */
import React from 'react'
import PropTypes from 'prop-types'

import Text from '@instructure/ui-elements/lib/components/Text'
import Checkbox from '@instructure/ui-forms/lib/components/Checkbox'
import Select from '@instructure/ui-forms/lib/components/Select'
import TextArea from '@instructure/ui-forms/lib/components/TextArea'
import I18n from 'i18n!edit_rubric'

import { assessmentShape } from './types'

const FreeFormComments = (props) => {
  const { savedComments, comments, saveLater, setComments, setSaveLater } = props
  const first = <option key="first" value="first">{I18n.t('[ Select ]')}</option>

  const options = savedComments.map((comment, ix) => (
    // eslint-disable-next-line react/no-array-index-key
    <option key={ix} value={ix.toString()}>{comment}</option>
  ))
  const selector = [
    (
      <Select
        key="select"
        label={I18n.t('Saved Comments')}
        assistiveText={I18n.t('Select from saved comments')}
        onChange={(_unused, el) => { setComments(savedComments[el.value])} }
      >
        {[first, ...options]}
      </Select>
    ),
    <br key="br" />
  ]

  return (
    <div className="edit-freeform-comments">
      {options.length > 0 ? selector : null}
      <TextArea
        label={I18n.t('Comments')}
        maxHeight="50rem"
        onChange={(e) => setComments(e.target.value)}
        resize="vertical"
        size="small"
        value={comments}
      />
      <br />
      <Checkbox
        checked={saveLater}
        label={I18n.t('Save this comment for reuse')}
        size="small"
        onChange={(event) => setSaveLater(event.target.checked)}
      />
    </div>
  )
}
FreeFormComments.propTypes = {
  savedComments: PropTypes.arrayOf(PropTypes.string).isRequired,
  comments: PropTypes.string,
  saveLater: PropTypes.bool,
  setComments: PropTypes.func.isRequired,
  setSaveLater: PropTypes.func.isRequired
}
FreeFormComments.defaultProps = {
  comments: '',
  saveLater: false
}

const commentElement = (assessment) => (
  assessment.comments_html ?
    // eslint-disable-next-line react/no-danger
    <div dangerouslySetInnerHTML={{ __html: assessment.comments_html }} />
    : assessment.comments
)

export const CommentText = ({ assessment, placeholder, weight }) => (
  <Text size="x-small" weight={weight}>
    {assessment !== null ? commentElement(assessment) : placeholder}
  </Text>
)
CommentText.propTypes = {
  assessment: PropTypes.shape(assessmentShape),
  placeholder: PropTypes.string,
  weight: PropTypes.string.isRequired
}
CommentText.defaultProps = {
  assessment: null,
  placeholder: ''
}

const Comments = (props) => {
  const { assessing, assessment, footer, ...commentProps } = props
  if (!assessing || assessment === null) {
    return (
      <div className="rubric-freeform">
        <CommentText
          assessment={assessment}
          placeholder={I18n.t("This area will be used by the assessor to leave comments related to this criterion.")}
          weight="normal"
        />
        {footer}
      </div>
    )
  }
  else {
    return (
      <FreeFormComments
        comments={assessment.comments}
        saveLater={assessment.saveCommentsForLater}
        {...commentProps}
      />
    )
  }
}
Comments.propTypes = {
  assessing: PropTypes.bool.isRequired,
  assessment: PropTypes.shape(assessmentShape),
  footer: PropTypes.node,
  savedComments: PropTypes.arrayOf(PropTypes.string).isRequired,
  setComments: PropTypes.func.isRequired,
  setSaveLater: PropTypes.func.isRequired
}
Comments.defaultProps = {
  footer: null
}

export default Comments
