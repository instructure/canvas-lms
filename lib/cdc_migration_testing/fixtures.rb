# frozen_string_literal: true

#
# Copyright (C) 2020 - present Instructure, Inc.
#
# This file is part of Canvas.
#
# Canvas is free software: you can redistribute it and/or modify it under
# the terms of the GNU Affero General Public License as published by the Free
# Software Foundation, version 3 of the License.
#
# Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
# A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
# details.
#
# You should have received a copy of the GNU Affero General Public License along
# with this program. If not, see <http://www.gnu.org/licenses/>.
#

Dir.glob(File.join(__dir__, "fixtures/**/*.rb")).each do |file|
  require_relative file
end

Dir.glob(File.join(__dir__, "../../gems/plugins/*/lib/cdc_migration_testing/fixtures/**/*.rb")).each do |file|
  require_relative file
end
