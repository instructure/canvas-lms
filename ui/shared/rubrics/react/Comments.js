/*
 * Copyright (C) 2018 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */
import React from 'react'
import PropTypes from 'prop-types'

import {ScreenReaderContent} from '@instructure/ui-a11y-content'
import {Text} from '@instructure/ui-text'
import {Checkbox} from '@instructure/ui-checkbox'
import {SimpleSelect} from '@instructure/ui-simple-select'
import {TextArea} from '@instructure/ui-text-area'
import I18n from 'i18n!edit_rubricComments'

import {assessmentShape} from './types'

const ellipsis = () => I18n.t('â€¦')

const truncate = comment => (comment.length > 100 ? comment.slice(0, 99) + ellipsis() : comment)

const slug = str => str.replace(/\W/g, '')

const FreeFormComments = props => {
  const {allowSaving, savedComments, comments, large, saveLater, setComments, setSaveLater} = props
  const first = (
    <SimpleSelect.Option key="first" id="first" value="first">
      {I18n.t('[ Select ]')}
    </SimpleSelect.Option>
  )

  const options = savedComments.map((comment, ix) => (
    <SimpleSelect.Option
      key={slug(comment).slice(-8)}
      id={`${slug(comment).slice(-6)}_${ix}`}
      value={ix.toString()}
      label={comment}
    >
      {truncate(comment)}
    </SimpleSelect.Option>
  ))
  const selector = [
    <SimpleSelect
      renderLabel={I18n.t('Saved Comments')}
      assistiveText={I18n.t('Select from saved comments')}
      onChange={(_unused, el) => {
        setComments(savedComments[el.value])
      }}
    >
      {[first, ...options]}
    </SimpleSelect>,
    <br key="br" />
  ]

  const saveBox = () => {
    if (allowSaving && large) {
      return (
        <Checkbox
          checked={saveLater}
          label={I18n.t('Save this comment for reuse')}
          size="small"
          onChange={event => setSaveLater(event.target.checked)}
        />
      )
    }
  }

  const label = I18n.t('Comments')
  const toScreenReader = el => <ScreenReaderContent>{el}</ScreenReaderContent>

  const commentClass = `edit-freeform-comments-${large ? 'large' : 'small'}`
  return (
    <div className={commentClass}>
      {options.length > 0 ? selector : null}
      <TextArea
        data-selenium="criterion_comments_text"
        label={large ? label : toScreenReader(label)}
        placeholder={large ? undefined : label}
        maxHeight="50rem"
        onChange={e => setComments(e.target.value)}
        resize="vertical"
        size="small"
        value={comments}
      />
      {large ? <br /> : null}
      {saveBox()}
    </div>
  )
}
FreeFormComments.propTypes = {
  allowSaving: PropTypes.bool.isRequired,
  savedComments: PropTypes.arrayOf(PropTypes.string).isRequired,
  comments: PropTypes.string,
  large: PropTypes.bool.isRequired,
  saveLater: PropTypes.bool,
  setComments: PropTypes.func.isRequired,
  setSaveLater: PropTypes.func.isRequired
}
FreeFormComments.defaultProps = {
  comments: '',
  saveLater: false
}

const commentElement = assessment => {
  if (assessment.comments_html || assessment.comments) {
    return (
      <div>
        <Text size="small" weight="bold">
          {I18n.t('Comments')}
        </Text>
        {assessment.comments_html ? (
          <div dangerouslySetInnerHTML={{__html: assessment.comments_html}} />
        ) : (
          <div>{assessment.comments}</div>
        )}
      </div>
    )
  } else {
    return null
  }
}

export const CommentText = ({assessment, placeholder, weight}) => (
  <span className="react-rubric-break-words">
    <Text size="x-small" weight={weight}>
      {assessment !== null ? commentElement(assessment) : placeholder}
    </Text>
  </span>
)
CommentText.propTypes = {
  assessment: PropTypes.shape(assessmentShape),
  placeholder: PropTypes.string,
  weight: PropTypes.string.isRequired
}
CommentText.defaultProps = {
  assessment: null,
  placeholder: ''
}

const Comments = props => {
  const {editing, assessment, footer, ...commentProps} = props
  if (!editing || assessment === null) {
    return (
      <div className="rubric-freeform">
        <CommentText
          assessment={assessment}
          placeholder={I18n.t(
            'This area will be used by the assessor to leave comments related to this criterion.'
          )}
          weight="normal"
        />
        {footer}
      </div>
    )
  } else {
    return (
      <FreeFormComments
        comments={assessment.comments}
        saveLater={assessment.saveCommentsForLater}
        {...commentProps}
      />
    )
  }
}
Comments.propTypes = {
  allowSaving: PropTypes.bool,
  editing: PropTypes.bool.isRequired,
  assessment: PropTypes.shape(assessmentShape),
  footer: PropTypes.node,
  large: PropTypes.bool,
  savedComments: PropTypes.arrayOf(PropTypes.string).isRequired,
  setComments: PropTypes.func.isRequired,
  setSaveLater: PropTypes.func.isRequired
}
Comments.defaultProps = {
  allowSaving: true,
  footer: null,
  large: true
}

export default Comments
