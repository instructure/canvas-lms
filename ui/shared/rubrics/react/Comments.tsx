/*
 * Copyright (C) 2018 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */
import React from 'react'
import type {ReactNode} from 'react'

import {ScreenReaderContent} from '@instructure/ui-a11y-content'
import {Text} from '@instructure/ui-text'
import {Checkbox} from '@instructure/ui-checkbox'
import {SimpleSelect} from '@instructure/ui-simple-select'
import {TextArea} from '@instructure/ui-text-area'
import {useScope as createI18nScope} from '@canvas/i18n'

const I18n = createI18nScope('edit_rubricComments')

const ellipsis = () => I18n.t('â€¦')

const truncate = (comment: string) =>
  comment.length > 100 ? comment.slice(0, 99) + ellipsis() : comment

const slug = (str: string) => str.replace(/\W/g, '')

interface Assessment {
  criterion_id: string
  comments?: string
  comments_html?: string
  points: {
    text?: string
    value?: number
    valid?: boolean
  }
  focusPoints?: number
  saveCommentsForLater?: boolean
}

interface FreeFormCommentsProps {
  allowSaving: boolean
  savedComments: string[]
  comments?: string
  large: boolean
  saveLater?: boolean
  setComments: (comments: string) => void
  setSaveLater: (saveLater: boolean) => void
}

const FreeFormComments = ({
  allowSaving,
  savedComments,
  comments = '',
  large,
  saveLater = false,
  setComments,
  setSaveLater,
}: FreeFormCommentsProps) => {
  const first = (
    <SimpleSelect.Option key="first" id="first" value="first">
      {I18n.t('[ Select ]')}
    </SimpleSelect.Option>
  )

  const options = savedComments.map((comment, ix) => (
    <SimpleSelect.Option
      key={slug(comment).slice(-8)}
      id={`${slug(comment).slice(-6)}_${ix}`}
      value={ix.toString()}
      label={comment}
    >
      {truncate(comment)}
    </SimpleSelect.Option>
  ))
  const selector = [
    <SimpleSelect
      renderLabel={I18n.t('Saved Comments')}
      assistiveText={I18n.t('Select from saved comments')}
      onChange={(_unused: any, el: any) => {
        setComments(savedComments[el.value])
      }}
      key="simple-select"
    >
      {[first, ...options]}
    </SimpleSelect>,
    <br key="br" />,
  ]

  const saveBox = () => {
    if (allowSaving && large) {
      return (
        <Checkbox
          checked={saveLater}
          label={I18n.t('Save this comment for reuse')}
          size="small"
          onChange={(event: React.ChangeEvent<HTMLInputElement>) =>
            setSaveLater(event.target.checked)
          }
        />
      )
    }
  }

  const label = I18n.t('Comments')
  const toScreenReader = (el: string) => <ScreenReaderContent>{el}</ScreenReaderContent>

  const commentClass = `rubric-comment edit-freeform-comments-${large ? 'large' : 'small'}`
  return (
    <div className={commentClass}>
      {options.length > 0 ? selector : null}
      <TextArea
        data-selenium="criterion_comments_text"
        label={large ? label : toScreenReader(label)}
        placeholder={large ? undefined : label}
        maxHeight="50rem"
        onChange={(e: React.ChangeEvent<HTMLTextAreaElement>) => setComments(e.target.value)}
        resize="vertical"
        size="small"
        value={comments}
      />
      {large ? <br /> : null}
      {saveBox()}
    </div>
  )
}

const commentElement = (assessment: Assessment) => {
  if (assessment.comments_html || assessment.comments) {
    return (
      <div>
        <Text size="small" weight="bold">
          {I18n.t('Comments')}
        </Text>
        {assessment.comments_html ? (
          <div dangerouslySetInnerHTML={{__html: assessment.comments_html}} />
        ) : (
          <div>{assessment.comments}</div>
        )}
      </div>
    )
  } else {
    return null
  }
}

interface CommentTextProps {
  assessment?: Assessment | null
  placeholder?: string
  weight: 'normal' | 'light' | 'bold'
}

export const CommentText = ({assessment = null, placeholder = '', weight}: CommentTextProps) => (
  <span className="react-rubric-break-words">
    <Text size="x-small" weight={weight}>
      {assessment !== null ? commentElement(assessment) : placeholder}
    </Text>
  </span>
)

interface CommentsProps {
  allowSaving?: boolean
  editing: boolean
  assessment?: Assessment | null
  footer?: ReactNode
  large?: boolean
  savedComments: string[]
  setComments: (comments: string) => void
  setSaveLater: (saveLater: boolean) => void
}

const Comments = ({
  allowSaving = true,
  editing,
  assessment,
  footer = null,
  large = true,
  savedComments,
  setComments,
  setSaveLater,
}: CommentsProps) => {
  if (!editing || assessment === null) {
    return (
      <div className="rubric-freeform">
        <CommentText
          assessment={assessment}
          placeholder={I18n.t(
            'This area will be used by the assessor to leave comments related to this criterion.',
          )}
          weight="normal"
        />
        {footer}
      </div>
    )
  } else {
    return (
      <FreeFormComments
        comments={assessment?.comments}
        saveLater={assessment?.saveCommentsForLater}
        allowSaving={allowSaving}
        savedComments={savedComments}
        large={large}
        setComments={setComments}
        setSaveLater={setSaveLater}
      />
    )
  }
}

export default Comments
