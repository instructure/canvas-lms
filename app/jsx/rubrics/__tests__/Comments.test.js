/*
 * Copyright (C) 2018 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */
import sinon from 'sinon'
import React from 'react'
import { shallow } from 'enzyme'
import Comments from '../Comments'

import { freeFormAssessment } from './fixtures'

describe('The Comments component', () => {
  const props = {
    assessing: true,
    assessment: freeFormAssessment.data[1],
    savedComments: [
      'I award you no points',
      'May god have mercy on your soul'
    ],
    saveLater: false,
    setComments: sinon.spy(),
    setSaveLater: sinon.spy()
  }

  const component = (mods) => shallow(<Comments {...{ ...props, ...mods }} />)
  const editor = (mods) => component(mods).find('FreeFormComments').shallow()
  const rating = (mods) => component(mods).find('CommentText').shallow()

  it('renders the root component as expected when assessing', () => {
    expect(component().debug()).toMatchSnapshot()
  })

  it('directly renders comments_html', () => {
    const el = rating({ assessing: false }).findWhere((e) => e.children().length === 0)
    expect(el.html()).toMatchSnapshot()
  })

  it('renders a placeholder when no assessment provided', () => {
    expect(rating({ assessing: false, assessment: null }).debug()).toMatchSnapshot()
  })

  it('shows no selector when no comments are presented', () => {
    expect(component({ savedComments: [] }).find('Select')).toHaveLength(0)
  })

  it('can used saved comments from before', () => {
    const setComments = sinon.spy()
    const el = editor({ setComments })
    const option = el.find('option').last()
    el.find('Select').prop('onChange')(null, { value: option.prop('value') })

    expect(setComments.args).toEqual([
      [option.text()]
    ])
  })

  it('can check / uncheck save for later', () => {
    const setSaveLater = sinon.spy()
    const el = editor({ setSaveLater })
    el.find('Checkbox').prop('onChange')({ target: { checked: true } })

    expect(setSaveLater.args).toEqual([[true]])
  })

  it('renders a footer after the comment when provided', () => {
    const el = component({ assessing: false, footer: <div>this is a footer</div> })

    expect(el.shallow().debug()).toMatchSnapshot()
  })
})
