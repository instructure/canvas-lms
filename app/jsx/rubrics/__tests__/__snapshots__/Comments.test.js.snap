// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Comments component directly renders comments_html 1`] = `"<div>I award you <b>no</b> points, and may God have mercy on your soul.</div>"`;

exports[`The Comments component renders a footer after the comment when provided 1`] = `
"<div className=\\"rubric-freeform\\">
  <CommentText assessment={{...}} placeholder=\\"This area will be used by the assessor to leave comments related to this criterion.\\" weight=\\"normal\\" />
  <div>
    this is a footer
  </div>
</div>"
`;

exports[`The Comments component renders a placeholder when no assessment provided 1`] = `
"<Text size=\\"x-small\\" weight=\\"normal\\" as=\\"span\\" letterSpacing=\\"normal\\">
  This area will be used by the assessor to leave comments related to this criterion.
</Text>"
`;

exports[`The Comments component renders the root component as expected when assessing 1`] = `"<FreeFormComments comments=\\"I award you no points, and may God have mercy on your soul.\\" saveLater={false} savedComments={{...}} setComments={[Function]} setSaveLater={[Function]} />"`;
